### PR TITLE
fix(camera): handle zero-radius Kannala-Brandt undistortion

### LINKED ISSUE #4308: W1b: `undistort_points_kannala_brandt` — the principal point returns `nan` in float16, because both singularity guards underflow
`undistort_points_kannala_brandt` returns `nan` for a point **at the principal point** when `params` is `float16`.

The function guards its two singular spots with hardcoded constants:

```python
th = rth.clamp(min=1e-16).sqrt()   # Newton start
...
denom = rth + eps                  # eps = 1e-8, final radial rescale
mag = radius_undistorted.abs() / denom
```

Both constants underflow to exactly `0.0` in `float16` (`1e-8` is below the smallest float16 subnormal, `~5.96e-8`; `1e-16` is far below it). The whole body runs in `params.dtype`, so with `float16` `params` the guards are no guards at all: at the principal point `rth == 0`, `denom == 0 + 0 == 0`, `radius_undistorted == tan(0) == 0`, and the result is `0.0 / 0.0 == nan` in both components.

`float32`, `float64` and `bfloat16` all return the exact origin — `bfloat16` survives because it keeps the float32 exponent range, so the underflow is specific to `float16`. The sibling `undistort_points_affine` has no such constants and returns the origin in every dtype.

## Repro

```python
import torch
from kornia.geometry.camera.distortion_affine import undistort_points_affine
from kornia.geometry.camera.distortion_kannala_brandt import undistort_points_kannala_brandt

for dtype in (torch.float64, torch.float32, torch.bfloat16, torch.float16):
    params = torch.tensor([100.0, 100.0, 4.0, 3.0, 0.1, 0.01, 0.001, 0.0001], dtype=dtype)
    principal_point = torch.tensor([4.0, 3.0], dtype=dtype)   # == (cx, cy)
    print(dtype, undistort_points_kannala_brandt(principal_point, params).tolist())

for dtype in (torch.float32, torch.float16):                  # the same on mps
    params = torch.tensor([100.0, 100.0, 4.0, 3.0, 0.1, 0.01, 0.001, 0.0001], dtype=dtype, device="mps")
    principal_point = torch.tensor([4.0, 3.0], dtype=dtype, device="mps")
    print("mps", dtype, undistort_points_kannala_brandt(principal_point, params).tolist())

# only params.dtype matters: the body casts the points to params.dtype and casts back at the end
params = torch.tensor([100.0, 100.0, 4.0, 3.0, 0.1, 0.01, 0.001, 0.0001], dtype=torch.float32)
print("f16 points / f32 params", undistort_points_kannala_brandt(torch.tensor([4.0, 3.0], dtype=torch.float16), params).tolist())

# the affine sibling, which has no hardcoded guards
for dtype in (torch.float64, torch.float32, torch.bfloat16, torch.float16):
    affine = torch.tensor([100.0, 100.0, 4.0, 3.0], dtype=dtype)
    print("affine", dtype, undistort_points_affine(torch.tensor([4.0, 3.0], dtype=dtype), affine).tolist())

print(torch.tensor(1e-8, dtype=torch.float16).item(), torch.tensor(1e-16, dtype=torch.float16).item())
```

## Observed

```
torch.float64 [0.0, 0.0]
torch.float32 [0.0, 0.0]
torch.bfloat16 [0.0, 0.0]
torch.float16 [nan, nan]
mps torch.float32 [0.0, 0.0]
mps torch.float16 [nan, nan]
f16 points / f32 params [0.0, 0.0]
affine torch.float64 [0.0, 0.0]
affine torch.float32 [0.0, 0.0]
affine torch.bfloat16 [0.0, 0.0]
affine torch.float16 [0.0, 0.0]
0.0 0.0
```

## Expected

The principal point undistorts to the origin in every supported dtype, as it already does in `float32`, `float64` and `bfloat16`. The two guards should be derived from the working dtype rather than hardcoded — `torch.finfo(dtype)`'s smallest **subnormal** (`torch.finfo(dtype).smallest_normal` is the smallest *normal* and would de-normalize the whole subnormal band, cf. #4162), or a dtype-scaled floor of the same spirit as the existing `1e-16` / `1e-8` in `float32`.

Note that the `1e-12` in the Gauss-Newton step denominator (`step = (thd - rth) / (d_thd + 1e-12)`) underflows in `float16` for the same reason; it is harmless at the principal point only because `d_thd` is `1.0` there, but it is the same class of defect.

## Classification

**Broken output**, dtype-scoped — a well-defined input silently producing `nan` in one dtype only. Idiom 2 (a hardcoded constant that does not inherit the input's dtype or precision).

## Disposition

n/a — single symbol.

## Environment

kornia @ `e8111a61` | torch 2.14.0 | python 3.11 | macOS | cpu and mps.

Found while writing the per-op `Convention:` block for this function in the camera conventions batch (#4262, sub-batch 5b). The block's guard bullet is scoped to the dtypes where the claim holds and points here; the claim is pinned by `test_convention_principal_point_undistorts_to_the_origin` and the current `float16` behavior by `test_wart_principal_point_undistorts_to_nan_in_float16_4308`, both in `tests/geometry/camera/test_distortion.py`. The batch documents and pins this behavior; it does not change it.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

### PR DESCRIPTION
**Pull request by victorwon2001**
## What does this pull request do?

`undistort_points_kannala_brandt` now handles an exact zero distorted radius with a structural mask instead of hardcoded epsilon guards. This makes the principal point return the exact origin in float16 and removes the additive `1e-8` radial-rescale bias that limited float64 round-trip precision, while preserving the existing 10 Gauss-Newton iterations and step denominator.

The convention documentation and regressions were updated accordingly, including mixed float16-points/float32-params behavior and representative float64 round trips.

## Related issue or discussion

Fixes #4308

## What did you check?

The focused principal-point and float64 precision regressions pass on CPU; the full float32/float64 camera-distortion module and both repository CPU half-precision profiles also pass with only their existing manifest xfails. MPS-specific cases were selected but skipped because this Linux host has no MPS device.

```text
$ pixi run --environment default test-module tests/geometry/camera/test_distortion.py --device=cpu --dtype=float16,float32,float64,bfloat16 -k 'principal_point_undistorts_to_the_origin or float64_round_trip_precision_4308'
$ pixi run --environment default test-module tests/geometry/camera/test_distortion.py --device=cpu --dtype=float32,float64
$ pixi run --environment default test-module tests/geometry/camera/test_distortion.py --known-failure-profile=cpu-float16 --xfail-known-failures
$ pixi run --environment default test-module tests/geometry/camera/test_distortion.py --known-failure-profile=cpu-bfloat16 --xfail-known-failures
$ pixi run --environment default typecheck
```

## How did AI help?

AI was used to inspect the affected implementation, callers, tests, and repository conventions; implement the zero-radius masking change; update the regression coverage and documentation; and run the checks above. The final diff was inspected for scope, compatibility, and unintended files.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
[P1] The zero-radius mask is computed after the radius has already underflowed. In float16, a representable non-principal point can therefore be classified as the principal point and silently collapsed to the origin. On `8132f75e`, using params `[100, 50, 4, 3, .1, .01, .001, .0001]` and the nearest representable point to the right of the centre, `[4.00390625, 3]`, the normalized x coordinate is `3.9041e-05`. Its square underflows to zero, `nonzero_radius` becomes false, and `undistort_points_kannala_brandt` returns `[0, 0]` instead of approximately `[3.9041e-05, 0]`.

Please derive zero-ness from the unsquared normalized components and compute the radius without half-precision underflow, then cover the nearest representable off-centre float16 point. Masking before the singular square root would also prevent exact-origin gradients from remaining NaN; the current patch returns a finite value there but still produces NaN point and parameter gradients. See `kornia/geometry/camera/distortion_kannala_brandt.py:166-172`.

[P2] This user-visible bug fix has no Unreleased changelog entry. Add one under Bug fixes citing #4308 and PR #4370.

The ordinary CPU float32/float64 distortion suite passes (53 passed, 1 skipped, 2 expected xfails), and hosted CI is green. Validation used the isolated PR checkout with its import path verified; the blocker is the uncovered near-origin behavior above.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).

**Review by ducha-aiki — APPROVED** [review]
Re-review of exact head 0a50ad763b9cbab3fa78c7d9620c820ff6b36966. The previous blockers are resolved: the zero-radius decision now uses the unsquared normalized components, float16 radius arithmetic is widened before squaring, the exact-origin path is shielded before the singular operations, and the release note cites both the issue and this PR. The new tests pin the nearest representable off-centre float16 point and the point/parameter Jacobians at the principal point.

Local validation in the isolated checkout verified the import origin and passed the 12 focused principal-point, near-origin, and float64 round-trip selections (8 passed, 4 dtype skips). Hosted ordinary precision, MPS, typecheck, docs, and pre-commit checks are green. The float16 job is still running, and the only completed test failure is the known XFeat inductor tolerance mismatch in an unrelated file, so CI still needs completion and a clean rerun before merge.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).
